File: Array/2. SecondLargest.ts
//1st Approach
function findSecondLargestElement(arr: number[]) {
  let newArr = [...arr].sort((a, b) => a - b);
  let largest = newArr[newArr.length - 1];
  let secondLargest = -1;

  for (let i = newArr.length - 2; i >= 0; i--) {
    if (newArr[i] < largest && newArr[i] > secondLargest) {
      secondLargest = newArr[i];
    }
  }
  return secondLargest;
}

// console.log(findSecondLargestElement([1, 2, 4, 7, 7, 5]));

//2nd Approach
function findSecondLargestElementSecondApproach(arr: number[]) {
  let largest = arr[0],
    sLargest = -1;
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] > largest) {
      largest = arr[i];
    }
  }

  for (let i = 0; i < arr.length; i++) {
    if (arr[i] < largest && arr[i] > sLargest) {
      sLargest = arr[i];
    }
  }

  return sLargest;
}

// console.log(findSecondLargestElementSecondApproach([1, 2, 4, 7, 7, 5]));

function findSecondLargestElementThirdApproach(arr: number[]) {
  let largest = arr[0],
    slargeest = -1;
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] > largest) {
      slargeest = largest;
      largest = arr[i];
    } else if (arr[i] < largest && arr[i] > slargeest) {
      slargeest = arr[i];
    }
  }
  console.log(largest);
  return slargeest; // returns the second largest element in the array. If there is no second largest, it returns -1.  // Example: findSecondLargestElementThirdApproach([1, 2, 4, 7, 7, 5]) => 7  // Example: findSecondLargestElementThirdApproach([1, 2, 3]) => -1  // Example: findSecondLargestElementThirdApproach([1])
}

// console.log(findSecondLargestElementThirdApproach([1, 2, 4, 7, 7, 5, 6]));
// console.log(findSecondLargestElementThirdApproach([10, 10]));
console.log(findSecondLargestElementThirdApproach([1]));

File: Array/2. SecondLargest.py
def bruteForce(arr):
    new_Arr = sorted(arr)
    n = len(new_Arr)
    largest = new_Arr[n-1]
    second_largest = -1
    for i in range(n-2, -1, -1):
        if largest > new_Arr[i]:
            second_largest = new_Arr[i]
            break
    return second_largest


# print(bruteForce([1, 2, 4, 7, 7, 5]))

def better_approach(arr):
    n = len(arr)
    largest = arr[0]
    second_largest = -1  # assuming list doesnot contain any negative element
    # if there would be any negative element, second_largest would be -Infinity
    for i in range(n):  # O(n)
        if (arr[i] > largest):
            largest = arr[i]

    for j in range(n):   # O(n)
        if arr[i] < largest and arr[i] > second_largest:
            second_largest = arr[i]

    return second_largest  # O(n+n) = O(2n)


# print(better_approach([1, 2, 4, 7, 7, 5]))

def optimal_approach(arr):
    n = len(arr)
    largest = arr[0]
    slargest = -1
    for i in range(n):
        if arr[i] > largest:
            slargest = largest
            largest = arr[i]
        elif arr[i] < largest and arr[i] > slargest:
            slargest = arr[i]

    return slargest


print(optimal_approach([1, 2, 4, 7, 7, 5]))
print(optimal_approach([1]))
